passes/pmgen/split_shiftx.pmg

```diff
@@ -0,0 +1,59 @@
+state <SigSpec> shiftxB
+
+match shiftx
+	select shiftx->type == $shiftx
+	select param(shiftx, \Y_WIDTH).as_int() > 1
+endmatch
+
+match macc
+	select macc->type == $macc
+	select param(macc, \B_WIDTH).as_int() == 0
+	optional
```

Comment:
Why is this optional? Below you only seem to do something when you found a macc cell.

Also this will run the code block below for all combinations of shiftx and macc cells inside the module. I.e. if you have 100 shiftx cells in a module and 100 macc cells, then the `code..endcode` block below is going to be executed 10000 for all the possible combinations, even if not a single macc is even connected to any shiftx cell.

So this should absolutely use something like `index <SigSpec> port(macc, \Y) === port(shiftx, \Y)` or `index <SigSpec> port(macc, \Y) === shiftxB`.

Reply:
> Why is this optional? Below you only seem to do something when you found a macc cell.

Not the full picture. I want to consider doing something for any multi-bit `$shiftx` where `Y_WIDTH > 1`, where:
1. If there is a `$macc` attached to B, check that it must conform to some conditions, otherwise ignore this `$shiftx` (i.e. let it get blown to soft logic). Accepted pattern: `$macc` `->` `$shiftx`.
2. If there is no `$macc` then accept the pattern `$shiftx`
3. Potentially other patterns down the line?

> Also this will run the code block below for all combinations of shiftx and macc cells inside the module. I.e. if you have 100 shiftx cells in a module and 100 macc cells, then the code..endcode block below is going to be executed 10000 for all the possible combinations, even if not a single macc is even connected to any shiftx cell.

Yup. I'm trying to nudge you towards thinking about (a) whether pmgen is the right tool for the job *ever*; (b) whether pmgen could be made into the right tool for the job. I think it would be a great little general purpose layer to have, without having to dive into writing your own custom pass to merge two cells. 

> So this should absolutely use something like index <SigSpec> port(macc, \Y) === port(shiftx, \Y) or index <SigSpec> port(macc, \Y) === shiftxB.

`index` doesn't currently let me examine subsets, right?
I've observed that sometimes `B_SIGNED` is set, and the MSB of `B` is `1'b0` -- thus checking if `macc.Y` === `shiftx.Y` will fail because of that extra sign bit. Thus, I used a `code` block to trim that sign bit off if the opportunity arises.

Reply:
My reasoning for this WIP is to have a stab at expressing what patterns I want to find in the design, and identify where the efficiencies are and how we can solve them. I think one of the things that all patterns would be able to benefit from would be the ability to walk connectivity (sing it with me) 

*The `$shiftx` cells is connected to the.... `$macc` cell.....*